scalardb-test/src/main/java/kelpie/scalardb/transfer/graphql/HttpTransferProcessor.java

```diff
@@ -0,0 +1,249 @@
+    super(config);
+
+    endpointUrl =
+        new URL("http", "localhost", (int) config.getUserLong("graphql_http", "port"), "/graphql");
+
+    this.numAccounts = (int) config.getUserLong("test_config", "num_accounts");
+    this.isVerification = config.getUserBoolean("test_config", "is_verification", false);
+    this.useCompactLog = config.getUserBoolean("test_config", "use_compact_log", true);
+  }
+
+  @Override
+  public void executeEach() throws Exception {
+    int fromId = ThreadLocalRandom.current().nextInt(numAccounts);
+    int toId = ThreadLocalRandom.current().nextInt(numAccounts);
+    int amount = ThreadLocalRandom.current().nextInt(1000) + 1;
+
+    logStart(fromId, toId, amount);
+
+    String txId;
+    try {
+      txId = transfer(fromId, toId, amount);
+    } catch (Exception e) {
+      logFailure(fromId, toId, amount, e);
+      throw e;
+    }
+    logSuccess(txId, fromId, toId, amount);
+  }
+
+  @Override
+  public void close() {}
+
+  private String transfer(int fromId, int toId, int amount) throws Exception {
+    int fromType = 0;
+    int toType = 0;
+    if (fromId == toId) {
+      toType = 1; // transfer between the same account
+    }
+
+    JsonObject variables1 =
+        Json.createObjectBuilder()
+            .add("from_id", fromId)
+            .add("from_type", fromType)
+            .add("to_id", toId)
+            .add("to_type", toType)
+            .build();
+    JsonObject response1 = sendRequest(GET_BALANCES, variables1);
+
+    JsonArray errors = response1.getJsonArray("errors");
+    if (errors != null) {
+      throw new GraphQlFailureException("", errors);
+    }
+
+    String txId =
+        response1.getJsonObject("extensions").getJsonObject("transaction").getString("id");
+    JsonObject data = response1.getJsonObject("data");
+    int fromBalance =
+        data.getJsonObject("from_account")
+            .getJsonObject(TransferCommon.TABLE)
+            .getInt(TransferCommon.BALANCE);
+    int toBalance =
+        data.getJsonObject("to_account")
+            .getJsonObject(TransferCommon.TABLE)
+            .getInt(TransferCommon.BALANCE);
+
+    JsonObject variables2 =
+        Json.createObjectBuilder()
+            .add("from_id", fromId)
+            .add("from_type", fromType)
+            .add("to_id", toId)
+            .add("to_type", toType)
+            .add("tx_id", txId)
+            .add("from_new_balance", fromBalance - amount)
+            .add("to_new_balance", toBalance - amount)
+            .build();
+    JsonObject response2 = sendRequest(PUT_BALANCES, variables2);
+
+    JsonArray errors2 = response2.getJsonArray("errors");
+    if (errors2 != null) {
+      throw new GraphQlFailureException(txId, errors2);
+    }
+
+    return txId;
+  }
+
+  private JsonObject sendRequest(String query, JsonObject variables) throws IOException {
+    HttpURLConnection connection = (HttpURLConnection) endpointUrl.openConnection();
```

Comment:
I'm not sure how much it affects the benchmark result, but I think we could use HTTP connection pooling. If we connect and disconnect every request, the three way handshake happens every request, which I think is inefficient.

Reply:
I believe the [`HttpURLConnection` is keep-alive transparently](https://docs.oracle.com/javase/8/docs/technotes/guides/net/http-keepalive.html), but maybe it should not be `disconnect()`ed. I'll do a quick check on this.

Reply:
I tested with a small HTTP client program that sends requests repeatedly to the local Nginx server.
It showed that regardless of whether or not `disconnect()` is called, one connection was established, which means the `Connection: keep-alive` response header was respected.
I checked it with the `netstat` command during the execution.

```console
$ netstat -p tcp -n | grep 8080
tcp4       0      0  127.0.0.1.8080         127.0.0.1.64409        ESTABLISHED
tcp4       0      0  127.0.0.1.64409        127.0.0.1.8080         ESTABLISHED
```

To disable keep-alive intentionally, it needed to give a system property `http.keepAlive=false`. With this property, the `netstat` showed many connections which means each request created a new connection. The `ESTABLISHED` state could not be observed because the connections didn't last long.

```console
$ netstat -p tcp -n | grep 8080
tcp4       0      0  127.0.0.1.8080         127.0.0.1.63544        TIME_WAIT
tcp4       0      0  127.0.0.1.8080         127.0.0.1.63545        TIME_WAIT
tcp4       0      0  127.0.0.1.8080         127.0.0.1.63546        TIME_WAIT
tcp4       0      0  127.0.0.1.8080         127.0.0.1.63547        TIME_WAIT
...
```

Besides this, I refactored the use of the try-with-resources statements in 82daeae.